accord-core/src/main/java/accord/impl/CommandTimeseries.java

```diff
@@ -0,0 +1,233 @@
+import static accord.local.SafeCommandStore.TestDep.WITH;
+import static accord.utils.Utils.ensureSortedImmutable;
+import static accord.utils.Utils.ensureSortedMutable;
+
+public class CommandTimeseries<D>
+{
+    public enum TestTimestamp
+    {BEFORE, AFTER}
+
+    private final Seekable keyOrRange;
+    protected final CommandLoader<D> loader;
+    public final ImmutableSortedMap<Timestamp, D> commands;
+    private final boolean ignoreTestKind;
+
+    public CommandTimeseries(Update<D> builder)
+    {
+        this.keyOrRange = builder.keyOrRange;
+        this.loader = builder.loader;
+        this.commands = ensureSortedImmutable(builder.commands);
+        this.ignoreTestKind = builder.ignoreTestKind;
+    }
+
+    CommandTimeseries(Seekable keyOrRange, CommandLoader<D> loader, ImmutableSortedMap<Timestamp, D> commands)
+    {
+        this.keyOrRange = keyOrRange;
+        this.loader = loader;
+        this.commands = commands;
+        this.ignoreTestKind = false;
+    }
+
+    public CommandTimeseries(Key keyOrRange, CommandLoader<D> loader)
+    {
+        this(keyOrRange, loader, ImmutableSortedMap.of());
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        CommandTimeseries<?> that = (CommandTimeseries<?>) o;
+        return keyOrRange.equals(that.keyOrRange) && loader.equals(that.loader) && commands.equals(that.commands);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        int hash = 1;
+        hash = 31 * hash + Objects.hashCode(keyOrRange);
+        hash = 31 * hash + Objects.hashCode(loader);
+        hash = 31 * hash + Objects.hashCode(commands);
+        return hash;
+    }
+
+    public D get(Timestamp key)
+    {
+        return commands.get(key);
+    }
+
+    public boolean isEmpty()
+    {
+        return commands.isEmpty();
+    }
+
+    /**
+     * All commands before/after (exclusive of) the given timestamp
+     * <p>
+     * Note that {@code testDep} applies only to commands that know at least proposed deps; if specified any
+     * commands that do not know any deps will be ignored.
+     * <p>
+     * TODO (expected, efficiency): TestDep should be asynchronous; data should not be kept memory-resident as only used for recovery
+     */
+    public <T> T mapReduce(SafeCommandStore.TestKind testKind, TestTimestamp testTimestamp, Timestamp timestamp,
+                           SafeCommandStore.TestDep testDep, @Nullable TxnId depId,
+                           @Nullable Status minStatus, @Nullable Status maxStatus,
+                           SafeCommandStore.CommandFunction<T, T> map, T initialValue, T terminalValue)
+    {
+
+        for (D data : (testTimestamp == TestTimestamp.BEFORE ? commands.headMap(timestamp, false) : commands.tailMap(timestamp, false)).values())
+        {
+            TxnId txnId = loader.txnId(data);
```

Comment:
I changed the ordering of this method due to cost to deserialize for `CommandsForKeys`, idea was to only deserialize what is needed and avoid touching deps until needed